src/graph.rs

```diff
@@ -3426,19 +3426,38 @@ impl<'a, 'graph> Builder<'a, 'graph> {
     package_req: &PackageReq,
   ) -> Option<Version> {
     // try to find in the list of existing versions first
-    if let Some(existing_versions) =
+    let versions = if let Some(existing_versions) =
       self.graph.packages.versions_by_name(&package_req.name)
     {
-      if let Some(version) = resolve_version(
-        &package_req.version_req,
-        existing_versions.iter().map(|nv| &nv.version),
-      ) {
-        return Some(version.clone());
-      }
+      existing_versions
+        .iter()
+        .map(|nv| &nv.version)
+        .collect::<Vec<_>>()
+    } else {
+      package_info.versions.keys().collect::<Vec<_>>()
+    };
+    let unyanked_versions = versions.clone().into_iter().filter(|v| {
+      package_info
+        .versions
+        .get(v)
+        .map(|i| !i.yanked)
+        .unwrap_or(true)
```

Comment:
We could avoid reaching into the hashmap for every version by instead only doing this if the version is resolved. See my comment below.

Reply:
> if the version is resolved

Do you mean unresolved? It won't catch lockfile entries then, since they are in `existing_versions` but not yielded from this function.